server/src-lib/Hasura/Server/Migrate.hs

```diff
@@ -84,6 +84,46 @@ data MigrationPair m = MigrationPair
   , mpDown :: Maybe (m ())
   }
 
+doesSchemaExist :: (MonadTx m) => SchemaName -> m Bool
+doesSchemaExist schemaName =
+  liftTx $ (runIdentity . Q.getRow) <$> Q.withQE defaultTxErrorHandler [Q.sql|
+    SELECT EXISTS
+    ( SELECT 1 FROM information_schema.schemata
+      WHERE schema_name = $1
+    ) |] (Identity schemaName) False
+
+
+doesTableExist :: (MonadTx m) => SchemaName -> TableName -> m Bool
+doesTableExist schemaName tableName =
+  liftTx $ (runIdentity . Q.getRow) <$> Q.withQE defaultTxErrorHandler [Q.sql|
+    SELECT EXISTS
+    ( SELECT 1 FROM pg_tables
+      WHERE schemaname = $1 AND tablename = $2
+    ) |] (schemaName, tableName) False
+
+isExtensionAvailable :: (MonadTx m) => SchemaName -> m Bool
+isExtensionAvailable schemaName =
+  liftTx $ (runIdentity . Q.getRow) <$> Q.withQE defaultTxErrorHandler [Q.sql|
+    SELECT EXISTS
+    ( SELECT 1 FROM pg_catalog.pg_available_extensions
+      WHERE name = $1
+    ) |] (Identity schemaName) False
```

Comment:
Did you have any particular reason for floating these out of `migrateCatalog`? It doesn’t seem like it was necessary to me, and having the definitions more local seems better than pulling them out.